### PR TITLE
Introducing a new FieldCaps ActionFilter 

### PR DESCRIPTION
**Pull request by thalurur**
*Issue #, if available:*
N/A

*Description of changes:*
Add a field caps action filter to return only formatted fields for rollup indices

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt [hunk 1]
```diff
@@ -0,0 +1,293 @@
+package com.amazon.opendistroforelasticsearch.indexmanagement.rollup.actionfilter
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt:1]: Missing license

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/IndexManagementPlugin.kt [hunk 2]
```diff
@@ -302,4 +304,8 @@ internal class IndexManagementPlugin : JobSchedulerExtension, NetworkPlugin, Act
     override fun getTransportInterceptors(namedWriteableRegistry: NamedWriteableRegistry, threadContext: ThreadContext): List<TransportInterceptor> {
         return listOf(rollupInterceptor)
     }
+
+    override fun getActionFilters(): MutableList<ActionFilter> {
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/IndexManagementPlugin.kt:308]: Did this need to be mutable list? Or can we change to List

**Review by qreshi — APPROVED** [review]

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/IndexManagementPlugin.kt (hunk 2 above)
Comment (thalurur) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/IndexManagementPlugin.kt:308]: good point, looks like I can just do List - will update

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt (hunk 1 above)
Comment (thalurur) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt:1]: right, will add

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+                val isRollupIndex = RollupSettings.ROLLUP_INDEX.get(clusterService.state().metadata.index(index).settings)
+                if (isRollupIndex) {
+                    rollupIndices.add(index)
+                } else {
+                    filteredIndices.add(index)
+                }
+            }
+
+            val hasRollupIndices = rollupIndices.size > 0
+            if (hasRollupIndices) {
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt:74]: I think if it doesn't have rollup we can just immediately call proceed right?
So something like:

```kotlin
if (rollupIndices.isEmpty()) {
    return chain.proceed(task, action, request, listener)
}
if (filteredIndices.isEmpty()) {
    val rewrittenResponse = rewriteResponse(mapOf(), arrayOf(), rollupIndices)
    return listener.onResponse(rewrittenResponse as Response)
}
request.indices(*filteredIndices.toTypedArray())
chain.proceed(task, action, request, object : ActionListener<Response> {
    override fun onResponse(response: Response) {
        logger.info("Has rollup indices will rewrite field caps response")
        response as FieldCapabilitiesResponse
        val reWrittenResponse = rewriteResponse(response.get(), response.indices, rollupIndices)
        listener.onResponse(reWrittenResponse as Response)
    }

    override fun onFailure(e: Exception) {
        listener.onFailure(e)
    }
})
```

i.e. no need to go through rest of logic as we don't need to do it.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+                val rewrittenResponse = rewriteResponse(mapOf(), arrayOf(), rollupIndices)
+                listener.onResponse(rewrittenResponse as Response)
+            } else {
+                chain.proceed(task, action, request, object : ActionListener<Response> {
+                    override fun onResponse(response: Response) {
+                        if (hasRollupIndices) {
+                            logger.info("Has rollup indices will rewrite field caps response")
+                            response as FieldCapabilitiesResponse
+                            val reWrittenResponse = rewriteResponse(response.get(), response.indices, rollupIndices)
+                            listener.onResponse(reWrittenResponse as Response)
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt:89]: nit: Here it's reWritten and above it's rewritten

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                    }
+                    fieldMappingsMap[fieldMapping]!!.add(rollupIndexFieldMappings.key)
+                }
+            }
+        }
+
+        return fieldMappingsMap
+    }
+
+    private fun getFieldType(fieldName: String, mappings: Map<*, *>): String? {
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt:196]: This seems very familiar to that other function we have somewhere in.. mappings service? We can extract that out into a common util method to reuse in both places?

**Comment by dbbaughe** [comment]
Mostly nits or small changes, can address now or later.

Biggest thing is let's start adding some comments to explain the purpose of each part. Can be hard to try and just get a high level view of what's happening (or supposed to happen) as need to read through every function to see what's going on. (This applies to PRs @qreshi and I have too). This can also be done as a follow up though.

**Review by dbbaughe — APPROVED** [review]

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt (hunk 4 above)
Comment (thalurur) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt:89]: makes sense, will update

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt (hunk 3 above)
Comment (thalurur) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt:74]: That looks cleaner, will update

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt (hunk 5 above)
Comment (thalurur) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt:196]: makes sense, will update

Comment (thalurur) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/actionfilter/FieldCapsFilter.kt:196]: moved common code to ``IndexUtils.getFieldFromMappings``

**Comment by thalurur** [comment]
> Mostly nits or small changes, can address now or later.
> 
> Biggest thing is let's start adding some comments to explain the purpose of each part. Can be hard to try and just get a high level view of what's happening (or supposed to happen) as need to read through every function to see what's going on. (This applies to PRs @qreshi and I have too). This can also be done as a follow up though.

Will do this as a follow up, will add for search interceptor as well in just one PR

**Review by dbbaughe — APPROVED** [review]

**Review by qreshi — APPROVED** [review]

**Pull request merged**